### PR TITLE
CORE-3102 Conditionally exclude GnuTLS

### PR DESCRIPTION
**Pull request by michael-redpanda**
Conditionally exclude GnuTLS from build when using OpenSSL

### CODE REVIEW

File: src/websocket/server.cc [hunk 1]
```diff
@@ -126,6 +143,28 @@ future<> connection::process() {
 
 static std::string sha1_base64(std::string_view source) {
     unsigned char hash[20];
+
+#ifdef SEASTAR_WITH_TLS_OSSL
+    const auto encode_capacity = [](size_t input_size) {
+        return (((4 * input_size) / 3) + 3) & ~0x3U;
```
Comment (graphcareful) [inline src/websocket/server.cc:149]: where does this maths come from?

File: src/websocket/server.cc [hunk 2]
```diff
@@ -27,11 +27,28 @@
...
+template<typename T, auto fn>
+struct ossl_deleter {
+    void operator()(T* ptr) { fn(ptr); }
+};
+
+template<typename T, auto fn>
+using ossl_handle = std::unique_ptr<T, ossl_deleter<T, fn>>;
+
+using evp_md_ptr = ossl_handle<EVP_MD, EVP_MD_free>;
+using evp_md_ctx_ptr = ossl_handle<EVP_MD_CTX, EVP_MD_CTX_free>;
```
Comment (graphcareful) [inline src/websocket/server.cc:49]: Could avoid this and just use a defer construct like whats going on in the existing tls stuff

File: include/seastar/net/tcp.hh [hunk 3]
```diff
@@ -2086,6 +2107,31 @@ tcp_seq tcp<InetTraits>::tcb::get_isn() {
     //   ISN = M + F(localip, localport, remoteip, remoteport, secretkey)
     //   M is the 4 microsecond timer
     using namespace std::chrono;
+#ifdef SEASTAR_WITH_TLS_OSSL
+    uint32_t hash[8];
```
Comment (graphcareful) [inline include/seastar/net/tcp.hh:2111]: Why did you choose a list of size 8?

**Review by graphcareful — COMMENTED** [review]
LGTM just left questions

File: src/websocket/server.cc (hunk 1 above)
Comment (michael-redpanda) [reply src/websocket/server.cc:149]: https://github.com/redpanda-data/redpanda/blob/52a68fcb0e3a1004c9a63334067344b1b20a47a7/src/v/utils/base64.cc#L21-L24

File: include/seastar/net/tcp.hh (hunk 3 above)
Comment (michael-redpanda) [reply include/seastar/net/tcp.hh:2111]: That's the size of a SHA256 hash.  It's twice as big as an MD5 digest

File: src/websocket/server.cc (hunk 2 above)
Comment (michael-redpanda) [reply src/websocket/server.cc:49]: yeah... I guess I could do that.

Comment (graphcareful) [reply src/websocket/server.cc:49]: up to the author though !

**Review by graphcareful — APPROVED** [review]

**Comment by michael-redpanda** [comment]
Force push `9ded57c`:

- Updates from PR review

**Comment by michael-redpanda** [comment]
Force push `1548f34`:

- Added back in websockets for OpenSSL build

File: include/seastar/net/tcp.hh [hunk 4]
```diff
@@ -2086,6 +2091,32 @@ tcp_seq tcp<InetTraits>::tcb::get_isn() {
...
+    unsigned int hash_size = sizeof(hash);
+
+    // Why SHA-256 for OpenSSL vs MD5?
+    // MD5 may be disabled if OpenSSL is in FIPS mode, also some bench testing
+    // has shown that the SHA-256 performance is equivalent or better than MD5
+    // as SHA256 is hardware accelerated on most modern CPU architectures
+    auto md_ptr = EVP_MD_fetch(nullptr, "SHA256", nullptr);
+    assert(md_ptr);
+    auto free_md_ptr = defer([&]() noexcept { EVP_MD_free(md_ptr); });
+    assert(hash_size = EVP_MD_get_size(md_ptr));
```
Comment (andijcr) [inline include/seastar/net/tcp.hh:2108]: was this meant to be `==`?
```suggestion
    assert(hash_size == EVP_MD_get_size(md_ptr));
```

Comment (michael-redpanda) [reply include/seastar/net/tcp.hh:2108]: great catch

File: tests/unit/tls_test.cc [hunk 5]
```diff
@@ -45,7 +45,9 @@
 #include "loopback_socket.hh"
 #include "tmpdir.hh"
 
+#ifndef SEASTAR_WITH_TLS_OSSL
 #include <gnutls/gnutls.h>
+#endif
```
Comment (andijcr) [inline tests/unit/tls_test.cc:50]: haven't looked at the rest of the file, but is the header needed at all?

File: CMakeLists.txt [hunk 6]
```diff
@@ -842,7 +842,7 @@ target_link_libraries (seastar
     # TODO(rob) - GnuTLS still a dependency until its last use (gnutls_hash_hd_t) in
     # seastar/net/tcp.hh can be removed
     # $<$<NOT:$<BOOL:${Seastar_WITH_OSSL}>>:GnuTLS::gnutls>
-    GnuTLS::gnutls
+    $<$<NOT:$<BOOL:${Seastar_WITH_OSSL}>>:GnuTLS::gnutls>
```
Comment (andijcr) [inline CMakeLists.txt:845]: nit: comment fix

File: tests/unit/tls_test.cc (hunk 5 above)
Comment (michael-redpanda) [reply tests/unit/tls_test.cc:50]: I believe this https://github.com/michael-redpanda/seastar/blob/1548f346ab0d350bbd9eef173083622eb7d0febd/tests/unit/tls_test.cc#L775 comes from that header.

**Comment by michael-redpanda** [comment]
Force push `78f315c`:

* Updates from PR

**Review by graphcareful — APPROVED** [review]

**Review by andijcr — APPROVED** [review]

**Pull request merged**